hpx/hpx_init_impl.hpp

```diff
@@ -84,10 +75,44 @@ namespace hpx
 #if defined(HPX_HAVE_CXX11_STD_QUICK_EXIT)
         std::at_quick_exit(detail::on_exit);
 #endif
+        if (params.rp_callback)
+        {
+            set_rp_callback(params.rp_callback);
+        }
 
-        return detail::run_or_start(f, desc_cmdline, argc, argv,
-            hpx_startup::user_main_config(cfg),
-            std::move(startup), std::move(shutdown), mode, true);
+        return detail::run_or_start(params.f, (*params.desc_cmdline_ptr),
+            params.argc, params.argv, hpx_startup::user_main_config(params.cfg),
+            std::move(params.startup), std::move(params.shutdown), params.mode,
+            params.rp_mode, true);
+    }
+
+    /// \brief Main entry point for launching the HPX runtime system.
+    ///
+    /// This is the main entry point for any HPX application. This function
+    /// (or one of its overloads below) should be called from the users `main()`
+    /// function. It will set up the HPX runtime environment and schedule the
+    /// function given by \p f as a HPX thread.
+    inline int init(
+        util::function_nonser<
+            int(hpx::program_options::variables_map& vm)
+        > const& f,
+        hpx::program_options::options_description const& desc_cmdline,
+        int argc, char** argv, std::vector<std::string> const& cfg,
+        startup_function_type startup, shutdown_function_type shutdown,
+        hpx::runtime_mode mode)
+    {
+        using hpx::program_options::options_description;
+        hpx::init_params iparams;
+        iparams.f = f;
+        iparams.desc_cmdline_ptr = std::make_shared<options_description>(
+            desc_cmdline);
```

Comment:
Can we assume that the user-supplied `options_description` object stays alive long enough such that we can just hold on to it by reference?

Reply:
I wanted to use a reference a first but the options_description is not furnishing a copy constructor nor a assignment operator (but for MSVC), so I still try to figure out what to do to remove it ^^

Reply:
Any ideas?

Reply:
Couldn't we capture the `options_descriptions` object by reference (potentially storing it as a `std::reference_wrapper<>`)?

Reply:
Thanks a lot ! :)